[BE] (build) 9 WEEK

seob7 opened the pull request.

seob7 committed:
feat:  회원가입 및 조회기능 구현

seob7 committed:
[BE] perf:  password 제거 및 Member 생성자 추가

- 요구 사항에 없던 password 제거
- Member 생성자를 사용한 setting

seob7 committed:
[BE] pref: grade type 을 enum으로 변경

seob7 committed:
[BE] perf: Map -> ConcurrentHashMap 으로 변경

Map 동시성 문제를 해결하기 위해
ConcurrentHashMap 사용

seob7 committed:
[BE] docs: readme.md 생성

seob7 committed:
[BE] perf: 객체 선언 수정

private static final 로 외부에서의 접근 및 중복 생성 방지

seob7 committed:
[BE] feat: product에 대한 getter setter 메소드 생성

seob7 committed:
[BE] feat: 상품 주문에 대한 main class 작성

seob7 committed:
[BE] feat: 상품 주문에 대한 ProductService class 작성

- order 메소드에서 member 의 아이디를 사용해 member의 등급을 조사한 뒤 할인율을 계산하였습니다.

seob7 committed:
[BE] feat: Repository 작성

- ProductRepositroy 인터페이스 작성
- ProductRepository를 구현한 MemeoryProductRepository 에서 메모리에 저장하는 기능 구현

seob7 committed:
[BE] docs: 상품 주문에 대한 기능 설명 추가

seob7 committed:
Merge branch 'FX-STUDY:main' into seob

seob7 committed:
Initial commit

seob7 committed:
[BE] (chore) gradle 로 변경

(1week code)순수자바에서 gradle 변경

seob7 committed:
[BE] (refactor) 1주차 코드 수정

- Rename MemoryMemberRepository to MemberRepository

- 자잘한 코드 수정

seob7 committed:
[BE] (chore) 클래스명 변경

MemberService 인터페이스를 작성하기 위해 기존 클래스명을 MemberServiceImpl로 변경

seob7 committed:
[BE] (feat) MemberService interface 작성

MemberService 인터페이스를 작성하고
MemberServiceImpl에 구현함

seob7 committed:
[BE] (refactor) 클래스명 및 자료형 수정

- Product -> Order
- price 와 discountedPrice에 대한 자료형 수정

seob7 committed:
[BE] (feat) 할인정책 구현

- 할인정책 interface 와 구현체 작성

seob7 committed:
[BE] (feat) orderService 구현

[BE] (feat) orderService 구현

주문서비스 인터페이스 및 구현체 작성

seob7 committed:
[BE] (refactor) Main class 수정

[BE] (refactor) Main class 수정

오타 수정 및 코드 리팩토링

seob7 committed:
[BE] (feat) RateDiscountPolicy 생성

[BE] (feat) RateDiscountPolicy 생성

할인을 30(%) 로 지정함
VIP 외 등급은 0을 리턴시킴

seob7 committed:
[BE] (docs) Add ISSUE

[BE] (docs) Add ISSUE

ISSUE 에 대한 내용 추가

seob7 committed:
[BE] (refactor) DiscountPolicy 클래스 리팩토링

[BE] (refactor) DiscountPolicy 클래스 리팩토링

- if문 수정
- discountRateAmount 정적변수로 변경
- discountFixAmount 정적변수로 변경

seob7 committed:
[BE] (refactor) 생성자를 통해 외부에서 주입받도록 변경

[BE] (refactor) 생성자를 통해 외부에서 주입받도록 변경

- 생성자를 사용해 FixDiscountPolicy 나 RateDiscoutPolicy를 주입할 수 있음.

seob7 committed:
[BE] (test) Main Test

[BE] (test) Main Test

orderServiceImpl 생성자 주입 테스트
(return 값으로 지정한 30이 출력됨)

seob7 committed:
[BE] (docs) Add solution

[BE] (docs) Add solution

1번 Issue에 대한 Solution 작성

seob7 committed:
[BE] (docs)  DI 이론 요약

[BE] (docs)  DI 이론 요약

seob7 committed:
[BE] (docs) Solution 수정

[BE] (docs) Solution 수정

seob7 committed:
[BE] (docs) 좋은 OOP 설계 5가지 원칙 정리

[BE] (docs) 좋은 OOP 설계 5가지 원칙 정리

SOLID
- SRP
- OCP
- LSP
- ISP
- DIP

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (docs) ISSUE 추가

[BE] (docs) ISSUE 추가

위반 내용 추가

seob7 committed:
[BE] (refactor) Change 'discountedPrice' to 'discountPrice'

[BE] (refactor) Change 'discountedPrice' to 'discountPrice'

seob7 committed:
[BE] (feat) calculatePrice 구현

[BE] (feat) calculatePrice 구현

OrderService구현체에 기능 구현.

seob7 committed:
[BE] (test) calculatePrice 확인

[BE] (test) calculatePrice 확인

입력 : 10000, VIP, 할인율 30%
출력 : 7000

seob7 committed:
[BE] (docs) Add ISSUE

[BE] (docs) Add ISSUE

seob7 committed:
[BE] (perf) calculatePrice 수정

[BE] (perf) calculatePrice 수정

RateDiscountPolicy 에서 return 값 변경으로 해결

seob7 committed:
[BE] (test) calculatePrice 테스트

[BE] (test) calculatePrice 테스트

seob7 committed:
[BE] (docs) Add solution

[BE] (docs) Add solution

calculatePrice에 대한 문제 해결

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (fix) 파일 전체 변경하여 오류 수정

[BE] (fix) 파일 전체 변경하여 오류 수정

이전 2week에서 Test 실행 오류로 인해 불가피하게 전체 파일을 변경하였습니다....

seob7 committed:
[BE] (build) Add lombok lib

[BE] (build) Add lombok lib

seob7 committed:
[BE] (chore) Member 클래스 수정

[BE] (chore) Member 클래스 수정

기존 Getter Setter 에서 스프링 어노테이션을 사용한 Getter Setter 로 변경

seob7 committed:
[BE] (chore) Order 클래스 수정

[BE] (chore) Order 클래스 수정

- toString 메소드 제거
- @ToString 등록

seob7 committed:
[BE] (chore) Map -> ConcurrentHashMap

[BE] (chore) Map -> ConcurrentHashMap

MemoryMemberRepository에 있는 Map을 동시성 이슈로 인한 ConcurrentHashMap로 변경

seob7 committed:
[BE] (chore) Order 클래스 수정

[BE] (chore) Order 클래스 수정

생성자를 자동 생성해주는 어노테이션 @AllArgsConstructor 추가

seob7 committed:
[BE] (docs) 이론요약 내용 추가

[BE] (docs) 이론요약 내용 추가

Spring IoC와 container에 대한 내용 추가

seob7 committed:
[BE] (docs) 이론요약 추가

[BE] (docs) 이론요약 추가

스프링 자동등록과 수동등록의 차이

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (bulid) 새 프로젝트 생성

[BE] (bulid) 새 프로젝트 생성

Project - Gradle-Groovy
Language - Java
Srping Boot - 3.2.0
Project Metadata
Group - hello
Artifact - item-service
Name - tiem-service
Package name - hello.itemservice
Packaging - Jar
Java - 17
Dependencies
Lombok
Spring Web
Thymeleaf

seob7 committed:
[BE] (feat) item domain

[BE] (feat) item domain

Getter Setter 로 된 item domain 추가

seob7 committed:
[BE] (feat) ItemRepository 생성

[BE] (feat) ItemRepository 생성

ItemRepository 및 MemoryItemRepository 생성하여 상품등록 및 상품상세 기능을 추가함

seob7 committed:
[BE] (feat) ItemService 생성

[BE] (feat) ItemService 생성

ItemService interface 생성

seob7 committed:
[BE] (feat) ItemServiceImpl 생성

[BE] (feat) ItemServiceImpl 생성

ItemService를 구현한 ItemServiceImpl 구현체 생성

seob7 committed:
[BE] (feat) ItemServiceTest 생성

[BE] (feat) ItemServiceTest 생성

seob7 committed:
[BE] (chore) public 제거

[BE] (chore) public 제거

seob7 committed:
[BE] (feat) AppConfig 생성

[BE] (feat) AppConfig 생성

seob7 committed:
[BE] (chore) ItemServiceImpl 수정

[BE] (chore) ItemServiceImpl 수정

이전에 까먹고 안넣은 implements 추가........

seob7 committed:
[BE] (chore) ItemServiceTest 수정

[BE] (chore) ItemServiceTest 수정

AppConfig를 사용하여 test 진행

seob7 committed:
[BE] (feat) findAll() 메서드 추가

[BE] (feat) findAll() 메서드 추가

ItemRepository에 모든 상품을 조회할 수 있는 findAll()메서드 추가

seob7 committed:
[BE] (feat) findAll() 테스트

[BE] (feat) findAll() 테스트

seob7 committed:
[BE] (chore) Rename add method

[BE] (chore) Rename add method

add -> addItem

seob7 committed:
[BE] (feat) addItemForm.html 생성

[BE] (feat) addItemForm.html 생성

상품등록 html 생성

seob7 committed:
[BE] (feat) home.html 생성

[BE] (feat) home.html 생성

main page 기능을 하는 home.html 생성

seob7 committed:
[BE] (feat) HomeController 클래스 생성

[BE] (feat) HomeController 클래스 생성

localhost 접속 시 home.html 로 이동

seob7 committed:
[BE] (feat) ItemController 클래스 생성

[BE] (feat) ItemController 클래스 생성

seob7 committed:
[BE] (feat) ItemForm 생성

[BE] (feat) ItemForm 생성

seob7 committed:
[BE] (refactor) 수동 -> 자동등록 변경

[BE] (refactor) 수동 -> 자동등록 변경

seob7 committed:
[BE] (feat) findAll() 메서드 수정

[BE] (feat) findAll() 메서드 수정

테스트 수정

seob7 committed:
[BE] (feat) HomeController 수정

[BE] (feat) HomeController 수정

itemService의 findAll 메서드를 사용해
Memory에 있는 값을 가져옴

seob7 committed:
[BE] (feat) home 수정

[BE] (feat) home 수정

Thymeleaf 템플릿 사용

seob7 committed:
[BE] (perf) item.itemId 추가

[BE] (perf) item.itemId 추가

seob7 committed:
[BE] (feat) 상품상세 Controller

[BE] (feat) 상품상세 Controller

seob7 committed:
[BE] (feat) 상품상세Form

[BE] (feat) 상품상세Form

seob7 committed:
[BE] (feat) Add editItem method to ItemRepository

[BE] (feat) Add editItem method to ItemRepository

상품 정보를 수정할 수 있는 기능 추가

seob7 committed:
[BE] (feat) Add editItem method to ItemService

[BE] (feat) Add editItem method to ItemService

상품 정보를 수정할 수 있는 기능 추가

seob7 committed:
[BE] (test) Add editItem method to ItemServiceTest

[BE] (test) Add editItem method to ItemServiceTest

editItem 메소드를 테스트할 수 있는 코드 작성

seob7 committed:
[BE] (test) Modify editItem test

[BE] (test) Modify editItem test

기존 테스트 코드 수정

seob7 committed:
[BE] (feat) Add editItemForm

[BE] (feat) Add editItemForm

상품 수정 form 추가

seob7 committed:
[BE] (feat) Add button to itemDetailForm

[BE] (feat) Add button to itemDetailForm

상품 수정 페이지로 넘어가는 버튼 추가

seob7 committed:
[BE] (perf) Modify editItem method to ItemService

[BE] (perf) Modify editItem method to ItemService

seob7 committed:
[BE] (perf) Modify editItem method to MemoryItemRepository

[BE] (perf) Modify editItem method to MemoryItemRepository

seob7 committed:
[BE] (feat) Add editItemForm controller to ItemController

[BE] (feat) Add editItemForm controller to ItemController

상품 수정과 관련된 기능 추가

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (build) Create new configuration

[BE] (build) Create new configuration

7week 작업을 위한 build 생성

seob7 committed:
[BE] (feat) Create NetworkClient

[BE] (feat) Create NetworkClient

seob7 committed:
[BE] (feat) Create BeanLifeCycleTest

[BE] (feat) Create BeanLifeCycleTest

seob7 committed:
[BE] (docs) 빈 생명주기 정리

[BE] (docs) 빈 생명주기 정리

seob7 committed:
[BE] (refactor) Refactor NetworkClient to implements Spring lifecycle callbacks

[BE] (refactor) Refactor NetworkClient to implements Spring lifecycle callbacks

InitializeBean, DisposableBean 인터페이스 구현

seob7 committed:
[BE] (docs) 인터페이스를 사용한 빈 생명주기 콜백

[BE] (docs) 인터페이스를 사용한 빈 생명주기 콜백

seob7 committed:
[BE] (refactor) Refactor NetworkClient lifecycle methods

[BE] (refactor) Refactor NetworkClient lifecycle methods

스프링 빈을 사용하기 위한 메서드 수정

seob7 committed:
[BE] (docs) 빈 등록 초기화, 소멸 메서드 정리

[BE] (docs) 빈 등록 초기화, 소멸 메서드 정리

seob7 committed:
[BE] (refactor) Refactor NetworkClient lifecycle methods

[BE] (refactor) Refactor NetworkClient lifecycle methods

 NetworkClient 클래스의 초기화와 소멸 메서드@PostConstruct와 @PreDestroy 를 사용하도록 함

seob7 committed:
[BE] (docs) 애노테이션 방법 정리

[BE] (docs) 애노테이션 방법 정리

@PostConstructor, @PreDestroy에 대한 내용 정리

seob7 committed:
[BE] (docs) 싱글톤 빈 스코프 정리

[BE] (docs) 싱글톤 빈 스코프 정리

seob7 committed:
[BE] (feat) Create SingletonTest

[BE] (feat) Create SingletonTest

seob7 committed:
[BE] (docs) 싱글톤 스코프 테스트 정리

[BE] (docs) 싱글톤 스코프 테스트 정리

seob7 committed:
[BE] (feat) Create PrototypeTest

[BE] (feat) Create PrototypeTest

seob7 committed:
[BE] (docs) 프로토타입 스코프 정리

[BE] (docs) 프로토타입 스코프 정리

싱글톤과 달리 프로토타입은 서로 다른 인스턴스를 생성하는 것을 확인할 수 있다.

seob7 committed:
[BE] (feat) Create SingletonWithPrototypeTest1.java

[BE] (feat) Create SingletonWithPrototypeTest1.java

프로토타입 빈 요청 테스트

seob7 committed:
[BE] (docs) 프로토타입 요청 정리

[BE] (docs) 프로토타입 요청 정리

요청 마다 항상 새로운 객체 인스턴스를 반환함

seob7 committed:
[BE] (refactor) Add singletonClientPrototype method to SingletonWithPrototypeTest1.java

[BE] (refactor) Add singletonClientPrototype method to SingletonWithPrototypeTest1.java

싱글톤 빈이 프로토타입 빈을 사용하도록 메서드 추가

seob7 committed:
[BE] (docs) 싱글톤 빈에서 프로토타입 빈 사용 정리

[BE] (docs) 싱글톤 빈에서 프로토타입 빈 사용 정리

seob7 committed:
[BE] (refactor) Edit Client static class to SingletonWithProrototyleTest1.java

[BE] (refactor) Edit Client static class to SingletonWithProrototyleTest1.java

싱글톤 빈이 프로토타입 빈을 사용할 때 마다 스프링 컨테이너에 새로 요청하도록 변경

seob7 committed:
[BE] (docs) DL에 대한 개념 정리

[BE] (docs) DL에 대한 개념 정리

seob7 committed:
[BE] (refactor) Using ObjectProvider to SingletonWithPrototypeTest1.java

[BE] (refactor) Using ObjectProvider to SingletonWithPrototypeTest1.java

ObjectProvider를 사용하도록 변경

seob7 committed:
[BE] (docs) ObjectProvider 정리

[BE] (docs) ObjectProvider 정리

seob7 committed:
[BE] (refactor) Using JSR-330 Provider to SingletonWithPrototypeTest1.java

[BE] (refactor) Using JSR-330 Provider to SingletonWithPrototypeTest1.java

기존 스프링의 ObjectProvider 에서 자바 표준 Object를 사용하도록 변경

seob7 committed:
[BE] (docs) JSR-330 Provider 정리

[BE] (docs) JSR-330 Provider 정리

seob7 committed:
[BE] (build) Add javax.inject dependency

[BE] (build) Add javax.inject dependency

seob7 committed:
[BE] (build) Add spring-boot-starter-web dependency

[BE] (build) Add spring-boot-starter-web dependency

seob7 committed:
[BE] (feat) Create MyLogger.java

[BE] (feat) Create MyLogger.java

seob7 committed:
[BE] (feat) Create LogDemoController.java

[BE] (feat) Create LogDemoController.java

ObjectProvider를 사용해 request scope 빈의 생성이 정상 처리되게 함

seob7 committed:
[BE] (feat) Create LogDemoService.java

[BE] (feat) Create LogDemoService.java

ObjectProvider를 사용해 request scope 빈의 생성이 정상 처리되게 함

seob7 committed:
[BE] (docs) request scope 정리

[BE] (docs) request scope 정리

seob7 committed:
[BE] (refactor) Refactor LogDemoService to proxyMode

[BE] (refactor) Refactor LogDemoService to proxyMode

프록시를 사용하기 위해 Provider 제거

seob7 committed:
[BE] (refactor) Refactor LogDemoController to proxyMode

[BE] (refactor) Refactor LogDemoController to proxyMode

프록시를 사용하기 위해 Provider 제거

seob7 committed:
[BE] (refactor) Refactor MyLogger to proxyMode

[BE] (refactor) Refactor MyLogger to proxyMode

프록시 설정

seob7 committed:
[BE] (docs) 프록시 정리

[BE] (docs) 프록시 정리

특징, 동작원리 정리

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (build) Create new configuration to week8

[BE] (build) Create new configuration to week8

seob7 committed:
[BE] (docs) 이미지 추가

[BE] (docs) 이미지 추가

seob7 committed:
[BE] (feat) 서블릿에 대한 코드

[BE] (feat) 서블릿에 대한 코드

reqeust, response 정리

seob7 committed:
[BE] (docs) 서블릿 정리

[BE] (docs) 서블릿 정리

seob7 committed:
[BE] (docs) 웹 애플리케이션 이해

[BE] (docs) 웹 애플리케이션 이해

seob7 committed:
[BE] (feat) Jsp/Servlet, MVC 패턴 적용

[BE] (feat) Jsp/Servlet, MVC 패턴 적용

seob7 committed:
[BE] (docs) 서블릿, JSP, MVC패턴 정리

[BE] (docs) 서블릿, JSP, MVC패턴 정리

seob7 committed:
[BE] (feat) Servlet MVC 프레임워크 생성

[BE] (feat) Servlet MVC 프레임워크 생성

seob7 committed:
[BE] (docs) MVC 프레임워크 만들기 정리

[BE] (docs) MVC 프레임워크 만들기 정리

seob7 committed:
Merge branch 'FX-STUDY:main' into seob

seob7 committed:
[BE] (build) 9주차 빌드

[BE] (build) 9주차 빌드

seob7 committed:
[BE] (refactor) 스프링을 사용하도록 함

[BE] (refactor) 스프링을 사용하도록 함

seob7 committed:
[BE] (docs) SECTION5 스프링MVC 정리

[BE] (docs) SECTION5 스프링MVC 정리

seob7 committed:
[BE] (build) SECTION6을 위한 프로젝트 생성

[BE] (build) SECTION6을 위한 프로젝트 생성

seob7 committed:
[BE] (feat) SECTION6 스프링 MVC 기본 기능

[BE] (feat) SECTION6 스프링 MVC 기본 기능

seob7 committed:
[BE] (docs) SECTION6 정리

[BE] (docs) SECTION6 정리

seob7 committed:
[BE] (build) SECTION7 item-service 빌드

[BE] (build) SECTION7 item-service 빌드

seob7 committed:
[BE] (feat) item-service 기능

[BE] (feat) item-service 기능

seob7 committed:
[BE] (docs) SECTION7 item-service 정리

[BE] (docs) SECTION7 item-service 정리

shinywoon commented:
좋습니다.
프로젝트하면서 활용할 수 있도록 완전한 숙지 바랍니다.

seob7 committed:
Merge branch 'FX-STUDY:main' into seob

seob7 committed:
[BE] (build) 10WEEK 초기 환경 구성

[BE] (build) 10WEEK 초기 환경 구성

seob7 committed:
[BE] (feat) Add Student Entity Class

seob7 committed:
[BE] (perf) Student 필드 타입 변경

seob7 committed:
[BE] (feat) 학생 관리 Repository CRUD 구현

seob7 committed:
[BE] (feat) 학생관리 Service 구현

seob7 committed:
[BE] (test) 학생관리 Service 단위별 테스트

seob7 committed:
[BE] (fix) stuMajort -> stuMajor 변경

seob7 committed:
[BE] (docs) 10WEEK docs

seob7 committed:
[BE] (refactor) 빌더 패턴 적용

seob7 committed:
[BE] (docs) 빌더패턴 적용

seob7 committed:
[BE] (fix) 정보 수정 Repository 수정

단일 학생 정보 수정 시 지정하지 않은 값에 대해 null이 발생하는 문제 발견
따라서 지정하지 않은 값에 기존 데이터를 참조함

seob7 committed:
[BE] (feat) studentController 추가

seob7 committed:
[BE] (docs) postman img

seob7 committed:
[BE] (docs) Controller Postman test 정리

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/Student.java:
```diff
@@ -0,0 +1,19 @@
+    private String stuName;
+    private String stuNum;
+    private String stuGrade;
+    private String stuMajort; //ENUM처리 고려
```
변수명 작명이 전반적으로 아쉽습니다.
짧으면 좋지만 짧은 이름 보다 명확한 이름이 더 좋은 이름입니다.
개발자는 누군가와 협업한다는 사실을 잊어서는 안됩니다.
`studentName` 처럼 조금 길더라도 명확한 변수명을 사용해주세요.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/Student.java:
```diff
@@ -0,0 +1,19 @@
+
+    private String stuName;
+    private String stuNum;
+    private String stuGrade;
```
Number 와 Grade 를 String type 으로 제공하기엔 제약조건이 많이 필요합니다.
Number 는 숫자로 이루어져야 하는 제약조건을 가져야하며,
Grade 는 1 ~ 5 라는 범위 제약 조건을 가져야 합니다.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/Student.java:
```diff
@@ -6,14 +6,16 @@
     private String stuName;
-    private String stuNum;
-    private String stuGrade;
+    private Long stuNum;
+    private int stuGrade;
```
바로 적용되어있네요...
하지만 부족한 제약조건에 대해 고민해주세요!
Number 에 Wrapper type 인 Long 을 적용하고
Grade 에 기본형인 int 를 선택한 이유가 있을까요?
Wrapper type 과 기본형의 차이는 무엇이 있을까요?

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/repository/StudentRepository.java:
```diff
@@ -0,0 +1,44 @@
+
+    public int countStu() {
+        return store.size();
+    }
```
로직은 전반적으로 원하는 기능을 수행하지만
예외 상황에 대한 고려가 필요합니다.
회원 가입하는 회원이 존재하는 경우?
삭제 중 문제가 생겼다면?

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final StudentRepository studentRepository;
+
+    public void signUp(Student stu) { //학생 입력
+        studentRepository.save(stu);
+    }
+
+    public Student showStudent(Long stuNum) { //단일 학생 조회
+        return studentRepository.findByStuNum(stuNum);
+    }
+
+    public List<Student> showAllStudent() { //모든 학생 조회
+        return studentRepository.findAll();
+    }
+
+    public void deleteStudent(Long stuNum) { // 단일 학생 삭제
+        studentRepository.delteByStuNum(stuNum);
+    }
+
+    public void deleteAllStudent() { // 다중 학생 삭제
+        studentRepository.deleteAll();
+    }
+
+    public Student updateStudent(Long stuNum, Student stu) { //단일 학생 정보 수정
+        return studentRepository.updateStu(stuNum, stu);
+    }
+
+    public int countAllStudent() {
+        return studentRepository.countStu();
+    }
```
Service Layer 가 존재하는 이유는 무엇일까요?
Repository 에서 하는 기능만 호출한다면 Service Layer 는 필요할까요?

shinywoon commented on 10 WEEK/studentManagement/src/test/java/fx/studentManagement/service/StudentServiceTest.java:
```diff
@@ -0,0 +1,147 @@
+        stu2.setStuName("KIM");
+        stu2.setStuGrade(4);
+        stu2.setStuMajort("SoftwareDept");
+        studentService.signUp(stu2);
```
`@BeforeEach` 를 잘 활용한 모습이지만 아쉬움이 남습니다.
`@PostConstructor` , `@PreConstructor` 등 Bean Life Cycle callBack 에 대해서 알아보세요

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/Student.java:
```diff
@@ -4,13 +4,15 @@
 @Getter
 @Setter
 @ToString
+@Builder
```
Builder PTN 과 Property 를 함께 사용하는 이유는 있을까요?

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/repository/StudentRepository.java:
```diff
@@ -35,6 +35,20 @@ public void deleteAll() {
     }
 
     public Student updateStu(Long stuNum, Student stu) {
+        Student existData = store.get(stuNum);
+        if(stu.getStuNum() == null)
+            stu.setStuNum(existData.getStuNum());
+        if (stu.getStuName() == null)
+            stu.setStuName(existData.getStuName());
+
+        Integer stuGrade = stu.getStuGrade();
+        if (stuGrade == null)
+            stu.setStuGrade(existData.getStuGrade());
+
+        if (stu.getStuMajor() == null)
+            stu.setStuMajor(existData.getStuMajor());
```
의도는 알겠으나... 좋은 아이디어는 아닌것 같습니다.
코드의 복잡성과 가독성이 많이 해지는것 같습니다.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -0,0 +1,77 @@
+        studentService.deleteStudent(stuNum);
+        return studentService.showAllStudent(); //확인을 위해 전체 출력
+    }
+
+    @DeleteMapping("/students") // 다중 학생 삭제
+    public List<Student> deleteAllStudent() {
+        studentService.deleteAllStudent();
+        return studentService.showAllStudent();
+    }
+
+    @PatchMapping("/students/{stuNum}") // 단일 학생 정보 수정
+    public Student updateStudent(@PathVariable Long stuNum, @RequestBody Student stu) {
+        studentService.updateStudent(stuNum, stu);
+        return studentService.showStudent(stuNum);
+    }
+
+    @GetMapping("/students/count") // 총 학생 조회
+    public int countStudent() {
+        return studentService.countAllStudent();
+    }
+
+    @PostConstruct
+    public void post() {
+        Student stu1 = Student.builder()
+                .stuNum(2100000L)
+                .stuName("PARK")
+                .stuGrade(2)
+                .stuMajor("SoftwareDept")
+                .build();
+        studentService.signUp(stu1);
+
+
+        Student stu2 = Student.builder()
+                .stuNum(1900000L)
+                .stuName("KIM")
+                .stuGrade(4)
+                .stuMajor("SoftwareDept")
+                .build();
+        studentService.signUp(stu2);
+    }
```
해당 Controller 의 resource 는 `student` 입니다.
하여 URL 에서 `/student` 가 중복으로 사용되는 것을 볼 수 있습니다.
Clean 코드의 기본 중 하나가 중복을 제거하자 입니다.
Class level 에 `@RequestMapping` 을 사용하면 공통 resource 를 관리 할 수 있습니다.

shinywoon reviewed (commented):
공부한것에 대해 정리도 잘하시고 아주 좋습니다.
리뷰에서 지적한 부분은 못했다는 뜻이 아닌 이 부분을 다음에는 고려해서 더 좋은 개발자가 되보자는 취지이니 꾸준히 열심히 해주시기 바랍니다.
공부하시다 힘든점, 배우고 싶은점이 있다면 언제든 연락 주시기바랍니다.

seob7 committed:
[BE] (refactor) Setter 제거, 변수명 변경

seob7 committed:
[BE] (refactor) 학생 등록 수정

- 예외처리 및 ResponseEntity 사용

seob7 committed:
[BE] (refactor) 단일학생조회 예외처리

seob7 committed:
[BE] (refactor) 학생 전체 조회 예외처리

seob7 committed:
[BE] (refactor) 단일 학생 삭제 예외처리

seob7 committed:
[BE] (refactor) 전체 학생 삭제 Controller 수정

seob7 committed:
[BE] (refactor) 학생 정보 수정 수정

- 예외처리 및 코드 간소화

seob7 committed:
[BE] (refactor) 변수명 변경

seob7 committed:
[BE] (refactor) 컨트롤러 수정

seob7 committed:
[BE] (test) 테스트케이스 수정

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/Student.java:
```diff
@@ -6,18 +6,16 @@
-    private String stuMajor; //ENUM처리 고려
+    private String studentName;
+    private Long studentNumber;
+    private Integer studentGrade;
```
int 와 Integer 차이를 알아야합니다.
Null 을 허용하지 않을 경우 int 를 사용하셔야 합니다.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -17,9 +17,13 @@ public class StudentController {
-    public List<Student> signUp(@RequestBody Student student) {
-        studentService.signUp(student);
-        return studentService.showAllStudent(); //학생 등록 확인을 위해 전체 출력
+    public ResponseEntity signUp(@RequestBody Student student) {
+        try {
+            studentService.signUp(student);
+            return new ResponseEntity<>("학생 등록 성공", HttpStatus.OK); // 학생 등록 확인을 위해 전체 출력
+        } catch (Exception e) {
+            return new ResponseEntity<>("이미 존재하는 학생입니다.", HttpStatus.BAD_REQUEST);
+        }
```
Try & Catch 는 Method 의 중요 법칙인 `한가지 일만 해라` 는 법칙을 위배합니다.
`한가지 일만 해라` 에 대해 공부해보세요.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -21,8 +23,10 @@ public void signUp(Student student) { //학생 입력
-        return studentRepository.findByStuNum(stuNum);
+    public Student showStudent(Long studentNumber) { //단일 학생 조회
+        if (studentRepository.findByStudentNumber(studentNumber) == null)
+            throw new RuntimeException("존재하지 않는 학생입니다.");
```
리터럴은 최대한 사용하지 않아야 합니다.
메세지 처리 부분을 공부해보세요.

shinywoon commented on 10 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -60,10 +60,15 @@ public ResponseEntity deleteAllStudent() {
-    public Student updateStudent(@PathVariable Long stuNum, @RequestBody Student stu) {
-        studentService.updateStudent(stuNum, stu);
-        return studentService.showStudent(stuNum);
+    @PatchMapping("/students/{studentNumber}") // 단일 학생 정보 수정
```
공통 Resource 는 Class level 에서 공통 처리 및 조합 하도록 설정하시기 바랍니다.

shinywoon reviewed (commented):
항상 발전하는 모습 너무 좋습니다.
항상 협업을 생각하시고 다른 사람들이 읽기 편한 프로그래밍에 신경써주도록 하세요

seob7 committed:
Merge remote-tracking branch 'upstream/main' into seob

seob7 committed:
[BE] (build) 12week 초기 빌드

seob7 committed:
[BE] (feat) 회원가입 및 정보수정 Form 구현

seob7 committed:
[BE] (feat) 학과 ENUM 처리

seob7 committed:
[BE] (perf) 학생등록 기능 개선

seob7 committed:
[BE] (pref) 학생 정보 필드 추가

seob7 committed:
[BE] (fix) 학생수정폼 오류수정

seob7 committed:
[BE] (fix) 날짜 오류 수정

seob7 committed:
[BE] (refactor) 학생정보수정 리팩토링

seob7 committed:
[BE] (build) 유효성 검사 에노테이션 빌드

seob7 committed:
[BE] (feat) enum Valid 구현

ENUM 유효성 검증을 위해 아래 내용 참고하였습니다.
https://jsy1110.github.io/2022/enum-class-validation/

seob7 committed:
[BE] (perf) LocalDate -> String 타입 변경으로 인한 수정

seob7 committed:
[BE] (feat) 학생 등록 유효성 검증 구현

seob7 committed:
[BE] (feat) NotFoundStudentException 구현

seob7 committed:
[BE] (feat) DuplicateStudentException 구현

seob7 committed:
[BE] (feat) ResponseMessage 구현

seob7 committed:
[BE] (feat) ExceptonAdvice 구현

seob7 committed:
[BE] (perf) 학생정보수정 유효성검증 구현

seob7 committed:
[BE] (perf) 예외처리 성능 개선

seob7 committed:
[BE] (refactor) 학과 ENUM 변경

seob7 committed:
[BE] (fix) Long Type Validation 수정

- 7자리 이하의 값에 유효성 검증 오류가 나오지 않았음
- @Range() 로 변경하여 오류 수정

seob7 committed:
[BE] (feat) ResponseMessage 추가

seob7 commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/form/SignUpForm.java:
```diff
@@ -5,14 +5,15 @@
-//    @Pattern(regexp = "\\d{7}") //숫자만 가능
-    @Digits(integer = 7, fraction = 0, message = "- 7자리, 숫자 ( 영문 , 특수 기호 금지 )")
+    //https://coderanch.com/t/586580/java/validate-length-number
+    @Range(min=7, max=7, message = "- 7자리, 숫자 ( 영문 , 특수 기호 금지 )")
```
@Range 도 안되네요....

shinywoon replied:
`@range` 는 범위입니다.
자릿수제한을 원하신다면 min=1000000 max=9999999 를 사용하면 가능하지 않을까요?

seob7 commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/enums/EnumValidator.java:
```diff
@@ -0,0 +1,31 @@
+import jakarta.validation.ConstraintValidatorContext;
+
+/*
+*  https://jsy1110.github.io/2022/enum-class-validation/
```
해당 내용 참고했습니다

shinywoon replied:
이렇게 복잡하게 안하셔도 Spring 은 `@Valid` , `@Validate` 라는 어노테이션을 제공합니다.

seob7 replied:
Enum 유효성 검사 에노테이션이 따로 존재하나요?

seob7 committed:
[BE] (fix) @Range 오류로 인한 주석처리

seob7 commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -83,4 +87,20 @@ private static Student saveStudent(SignUpForm signUpForm) {
+        Student student = Student.builder()
+                .studentNumber(existsStudent.getStudentNumber())
+                .studentGrade(existsStudent.getStudentGrade())
+                .studentName(editStudentForm.getStudentName())
```
`EditStudentForm` 으로 사용자의 새로운 정보를 기존 데이터에 덮어 씌울 때 `Student` 폼을 사용하지 않으면 나머지는 null이 되는 문제가 있어 이러한 방식을 사용했는데 맞는지 모르겠네요...

shinywoon replied:
form 을 Entity로 변환하는 과정은 실무에서 사용하는 방법입니다.
직접 개발하는 경험을 더 해보신다면 보통 entity 객체의 field와 form 객체의 field 가 100% 일치하지 않는일이 허다하다는 사실을 알게 될 것 입니다.

seob7 committed:
[BE] (refactor) Controller 수정

seob7 committed:
[BE] (refactor) 학생 정보 수정 메서드명 변경

updateStudent -> editStudentInformation

seob7 committed:
[BE] (docs) 12WEEK 정리

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -18,9 +20,9 @@ public class StudentController {
+            studentService.signUp(signUpForm);
             return new ResponseEntity<>("학생 등록 성공", HttpStatus.OK); // 학생 등록 확인을 위해 전체 출력
         } catch (Exception e) {
             return new ResponseEntity<>("이미 존재하는 학생입니다.", HttpStatus.BAD_REQUEST);
```
리터럴 처리에 신경써주세요.
만일 동일한 메세지가 100곳에서 사용된다면 100곳을 수정하는 대참사가 일어납니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -15,14 +17,17 @@ public class StudentService {
+    public void signUp(SignUpForm signUpForm) { //학생 입력
 
-        if(studentRepository.findByStudentNumber(student.getStudentNumber()) != null)
+        if(studentRepository.findByStudentNumber(signUpForm.getStudentNumber()) != null)
```
if 안에 들어간 조건문이 읽기 쉽지 않습니다.
가독성을 어떻게하면 개선할 수 있는지 고민해보시기 바랍니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -15,14 +17,17 @@ public class StudentService {
+        if(studentRepository.findByStudentNumber(signUpForm.getStudentNumber()) != null)
             throw new RuntimeException("이미 존재하는 학생입니다.");
 
+        Student student = saveStudent(signUpForm);
```
메소드명은 `saveStudent` 이지만 실제로 수행하는 일은 signUpForm을 Student instance 로 변환하는 일입니다.
메소드명에서 메소드가 하는 일이 짐작가능해야합니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -66,9 +66,9 @@ public int countAllStudent() {
-        int day = signUpForm.getStudentBirth().getDay(); //여기서 약간 계산오류 발생. 변경하기!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+        int year = signUpForm.getStudentBirth().getYear();
+        int month = signUpForm.getStudentBirth().getMonthValue();
+        int day = signUpForm.getStudentBirth().getDayOfMonth();
```
변수명이 일관성없게 사용되고 있습니다.
Year 값은 year, Month 는 MonthValue ..., DayOfMonth 까지.... 일관성이 중요합니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -64,9 +64,9 @@ public ResponseEntity deleteAllStudent() {
 
     @PatchMapping("/{studentNumber}") // 단일 학생 정보 수정
-    public ResponseEntity updateStudent(@PathVariable Long studentNumber, @RequestBody Student student) {
+    public ResponseEntity updateStudent(@PathVariable Long studentNumber, @RequestBody EditStudentForm editStudentForm) {
```
EditStudentForm 객체에 studentNumber field 를 넣어 pathVariable 을 없애 가독성을 더 높히는 방법이 어떨까요?

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -50,9 +50,13 @@ public void deleteAllStudent() { // 다중 학생 삭제
-    public Student updateStudent(Long studentNumber, Student student) { //단일 학생 정보 수정
-        if(studentRepository.findByStudentNumber(studentNumber) == null)
+    public Student updateStudent(Long studentNumber, EditStudentForm editStudentForm) { //단일 학생 정보 수정
+        Student existsStudent = studentRepository.findByStudentNumber(studentNumber);
+        if(existsStudent == null)
```
existstStudent -> Boolean 을 반환할 것으로 예상되나 반환값이 Student 인 것이 명확성이 떨어진다는 것을 반증합니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/service/StudentService.java:
```diff
@@ -83,4 +87,20 @@ private static Student saveStudent(SignUpForm signUpForm) {
     }
+
+
+    private static Student editStudentInfo(EditStudentForm editStudentForm, Student existsStudent) {
```
`editStudentInfo` 라는 메소드명을 사용했지만 실제 동작은 EditForm 을 Student 객체로 변환하는 행위입니다.

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/entity/enums/Major.java:
```diff
@@ -1,20 +1,9 @@
-    SoftwareDept("소프트웨어학과"),
-    ComputerEngineeringDept("컴퓨터공학과"),
-    IndustrialDesignDept("산업디자인학과");
+    소프트웨어학과,
+    컴퓨터공학과,
+    산업디자인학과
```
????

seob7 replied:
```
{
    "studentNumber" : 2026055,
    "studentName" : "이동섭",
    "studentBirth" : "2001_09_07",
    "studentMajor" : "소프트웨어학과",
    "studentSemester" : 2,
    "studentAddress" : "충주시",
    "studentGrade" : 1
}
```
studentMajor에 한글로 "소프트웨어학과" 같은 요청이 오면 SoftwareDept 로 영문으로 매핑하는 코드를 따로 작성해야 하는건가요?...?

shinywoon commented on 12 WEEK/studentManagement/src/main/java/fx/studentManagement/controller/StudentController.java:
```diff
@@ -21,59 +23,45 @@ public class StudentController {
+    public ResponseEntity signUp(@Valid @RequestBody SignUpForm signUpForm, BindingResult bindingResult) {
+
+        if(bindingResult.hasErrors())
+            return new ResponseEntity<>(bindingResult.getAllErrors(), HttpStatus.BAD_REQUEST);
```
responseForm 을 사용했던거 같은데 .... ResponseEntity<> 를 사용하는 이유가 있을까요?

shinywoon reviewed (commented):
중간중간 코드가 기술적으로 아쉬운부분들이 많습니다.
보완하신다면 좋은 개발자가 될 수 있을것 같습니다.

seob7 committed:
[BE] (fix) 학번 자릿수 제한 수정

seob7 committed:
[BE] (refactor) 학생 정보 수정 refactoring

- 폼데이터 추가
- 메서드명 변경

seob7 committed:
[BE] (refactor) 메서드명 변경

- saveStudent -> convertSignUpFormToStudent

seob7 committed:
[BE] (perf) 학생 존재 여부 로직 개선

- existsStudent 메서드로 대체하여 가독성과 일관성을 향상시킴

seob7 committed:
[BE] (reafctor) ENUM 변경

seob7 committed:
[BE] (perf) 빌더패턴 적용

seob7 committed:
[BE] (test) 테스트 코드 변경
